lib/mongo/server/connection_pool.rb

```diff
@@ -321,7 +322,9 @@ def check_out(service_id: nil)
                   next
                 end
 
-                if connection.generation != generation(service_id: connection.service_id)
+                if connection.generation != generation(
+                  service_id: connection.service_id
+                )
                   # Stale connections should be disconnected in the clear
                   # method, but if any don't, check again here
```

Comment:
In my understanding this operation shouldn't be done if the connection is pinned. The language in https://github.com/mongodb/specifications/blob/master/source/load-balancers/load-balancers.rst#connection-pinning and subsequent sections says that connection is not returned to the pool while it is pinned, which in effect is how it avoids nuking the connection if (hopefully this is correct) an election happens while a cursor is being iterated. I think master is already somewhat deficient in this manner in that it might close a connection for a service leaving operations for that service to fail in the future, and I'm not sure whether master is already deficient for cursors.

What I think needs to happen is: 1) the connection needs to know that something is pinned to it (session or cursor), 2) while anything is pinned to connection it shouldn't be closed due to generation becoming stale, max idle time being exceeded, etc.

In theory every connection that something is pinned to should be used eventually for an operation which would unpin the connection (killcursors/aborttransaction for example), and when that operation completes and the connection is checked in it could be closed, but the unpinning logic might need to be added e.g. for the cursors.

Reply:
Created https://jira.mongodb.org/browse/RUBY-2936